lib/FrontendTool/FrontendTool.cpp

```diff
@@ -262,6 +262,251 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+      return;
+    }
+    if (::isClangOverlayOf(module, reexport)) {
+      // For overlays, we need to have a dependency on the underlying module.
+      // Otherwise, we'll accidentally create a Swift -> Swift cycle.
+      abiExportMap[module].insert(
+        reexport->getTopLevelModule(/*preferOverlay*/false));
+      return;
+    }
+    abiExportMap[module].insert(
+        reexport->getTopLevelModule(/*preferOverlay*/true));
+  }
+
+  /// Mark \p importedModule as exposed for \p module's ABI.
+  void reexposeImportedABI(ModuleDecl *module, ModuleDecl *importedModule) {
+    if (module == importedModule) {
+      crashOnInvariantViolation([&](llvm::raw_string_ostream &os) {
+        os << "module ";
+        printModule(module, os);
+        os << " imports itself!\n";
+      });
+      return;
+    }
+    addToABIExportMap(module, importedModule);
+    for (auto reexportedModule: abiExportMap[importedModule])
+      addToABIExportMap(module, reexportedModule);
+  }
+
+  /// Recursive step in computing ABI dependencies.
+  ///
+  /// Use this method instead of using the \c forClangModule/\c forSwiftModule
+  /// methods.
+  void computeABIDependenciesForModule(ModuleDecl *module) {
+    if (llvm::find(searchStack, module) != searchStack.end()) {
+      // If we have a Clang module, then stop recursing, because Clang modules
+      // can have weird dependencies on each other. For example, traditionally
+      // modules depend on (re-export) their submodules, but in practice,
+      // submodules can also depend on their parents (or generally any ancestor)
+      // or other submodules. This can lead to intra-top-level-module cycles.
+      if (module->isNonSwiftModule()) {
+        return;
+      }
+      crashOnInvariantViolation([&](llvm::raw_string_ostream &os) {
+        os << "unexpected cycle in import graph!\n";
+        for (auto mod: searchStack) {
+          printModule(mod, os); os << "\ndepends on ";
+        }
+        printModule(module, os); os << '\n';
+      });
+      return;
+    }
+    if (::contains(visited, module))
+      return;
+    searchStack.push_back(module);
+    if (module->isNonSwiftModule())
+      computeABIDependenciesForClangModule(module);
+    else
+      computeABIDependenciesForSwiftModule(module);
+    searchStack.pop_back();
+    visited.insert(module);
+  }
+
+  void computeABIDependenciesForClangModule(ModuleDecl *module) {
+    SmallPtrSet<ModuleDecl *, 32> imports;
+    ::getImmediateImports(module, imports);
+    for (auto import: imports) {
+      // FIXME: [Varun] This hack avoids some 2-module cycles which don't
+      // seem to have any clear pattern to them...
+      auto moduleName = module->getName().str();
+      auto importName = import->getName().str();
+      if ((moduleName == "SwiftShims" && importName == "Swift")
+          || (moduleName == "SwiftOverlayShims" && importName == "Darwin")) {
+        continue;
+      }
```

Comment:
@brentdax any ideas on why this might be needed? Otherwise, I end up seeing dependency cycles like:

```
Swift (Swift) -> SwiftShims (Clang) -> Swift (Swift)
Darwin (Swift) -> SwiftOverlayShims (Clang) -> Darwin (Swift)
```

I'm not really sure what the commonality is.

Reply:
Fixed this after discussion with Robert and Hamish. 😄